### PR TITLE
[AutoMM] Enforce valid metric names and add fallback metrics for each problem type

### PR DESCRIPTION
**Pull request by FANGAreNotGnu**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 1]
```diff
@@ -75,61 +81,60 @@ def infer_metrics(
...
-    elif problem_type is None and is_matching:
-        return RECALL, NDCG
     else:
-        raise NotImplementedError(f"Problem type: {problem_type} is not supported yet!")
-
-    validation_metric_name = eval_metric_name
+        eval_metric_name = EVALUATION_METRICS_FALLBACK[problem_type]
+        validation_metric_name = VALIDATION_METRICS_FALLBACK[problem_type]
 
     return validation_metric_name, eval_metric_name
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:139]: The current logic ignores the `validation_metric_name` from `infer_metrics`'s input. Do we still support users to customize validation metric? I think previously it's mainly used by detection for the map issue.

File: multimodal/src/autogluon/multimodal/constants.py [hunk 2]
```diff
@@ -162,6 +165,59 @@
...
+    IMAGE_TEXT_SIMILARITY: [None],  # use matching metrics
+    FEW_SHOT_CLASSIFICATION: METRICS[BINARY].keys() + METRICS[MULTICLASS].keys(),
+    SEMANTIC_SEGMENTATION: [BINARY_IOU],  # TODO: ADD METRICS FOR SEMANTIC_SEGMENTATION
+}
+EVALUATION_METRICS_FALLBACK = {problem_type: metrics[0] for problem_type, metrics in EVALUATION_METRICS.items()}
+EVALUATION_METRICS_FALLBACK[BINARY] = ROC_AUC
+EVALUATION_METRICS_FALLBACK[MULTICLASS] = ACCURACY
+EVALUATION_METRICS_FALLBACK[REGRESSION] = RMSE
+EVALUATION_METRICS_FALLBACK[FEW_SHOT_CLASSIFICATION] = ACCURACY
+VALIDATION_METRICS = {
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/constants.py:193]: We can replace previous `VALID_METRICS` with `VALIDATION_METRICS`. No need to keep them both.

**Review by zhiqiangdon — COMMENTED** [review]
https://github.com/autogluon/autogluon/issues/3414#issuecomment-1791355728
This PR should fix the f1 related issues for both binary and multiclass. Also, check the this function (https://github.com/autogluon/autogluon/blob/master/multimodal/src/autogluon/multimodal/optimization/utils.py#L234) to make sure f1 metrics are created correctly for different f1 variants and problem types.

**Review by zhiqiangdon — COMMENTED** [review]
Need to add tests for f1 related metrics for both binary and multiclass. Predictor fitting should work without errors. Also, test `get_metric` to make sure the returned metric function can compute values as expected.

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 1 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:139]: We can ignore for now. But we might need to support this later.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 3]
```diff
@@ -74,62 +75,64 @@ def infer_metrics(
...
-            UserWarning,
-        )
+    if eval_metric_name is not None:
+        if eval_metric_name.lower() not in problem_property.supported_evaluation_metrics:
+            warnings.warn(
+                f"Metric {eval_metric_name} is not supported as the evaluation metric for {problem_type}. "
+                f"The evaluation metric is changed to {problem_property.fallback_evaluation_metric} by default."
+            )
+            if problem_property.fallback_evaluation_metric is not None:
+                eval_metric_name = problem_property.fallback_evaluation_metric
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:119]: What's the validation metric in this case?

File: multimodal/tests/unittests/others/test_metrics.py
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others/test_metrics.py:1]: Consider adding `test_infer_metrics` for the utile function `infer_metrics`. In the tests, we can cover all problem types and some invalid evaluation metrics.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 4]
```diff
@@ -74,62 +75,64 @@ def infer_metrics(
...
-            f", ROC-AUC for binary classification problem, and RMSE for regression problems.",
-            UserWarning,
-        )
+    if eval_metric_name is not None:
+        if eval_metric_name.lower() not in problem_property.supported_evaluation_metrics:
+            warnings.warn(
+                f"Metric {eval_metric_name} is not supported as the evaluation metric for {problem_type}. "
+                f"The evaluation metric is changed to {problem_property.fallback_evaluation_metric} by default."
+            )
+            if problem_property.fallback_evaluation_metric is not None:
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:118]: Currently `SEMANTIC_SEGMENTATION`, `CLASSIFICATION`, matching, `FEATURE_EXTRACTION`, and`ZERO_SHOT_IMAGE_CLASSIFICATION` does not need `fallback_evaluation_metric`. Later we should try to integrate `SEMANTIC_SEGMENTATION`'s metric here

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 5]
```diff
@@ -74,62 +75,64 @@ def infer_metrics(
...
-            ROC_AUC,
-            AVERAGE_PRECISION,
-            F1,
-        ]:
-            raise ValueError(f"Metric {eval_metric_name} is only supported for binary classification.")
-
-        if eval_metric_name in VALID_METRICS:
-            validation_metric_name = eval_metric_name
+        if problem_type is None:
+            eval_metric_name, validation_metric_name = MATCHING_METRICS_WITHOUT_PROBLEM_TYPE[problem_type]
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:96]: Just fixed the typo, should be `eval_metric_name, validation_metric_name = MATCHING_METRICS_WITHOUT_PROBLEM_TYPE`

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 3 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:119]: It's inferred below

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 6]
```diff
@@ -74,62 +75,64 @@ def infer_metrics(
...
-        elif validation_metric_name.lower() == DIRECT_LOSS:
-            return DIRECT_LOSS, MAP
+        if eval_metric_name.lower() in problem_property.supported_validation_metrics:
+            logger.info(f"Metric {eval_metric_name} is used as the validation metric. ")
+            validation_metric_name = eval_metric_name
         else:
-            raise ValueError(
-                f"Problem type: {problem_type}, validation_metric_name: {validation_metric_name} is not supported!"
+            # TODO: change semantic segmentation logics to use this function to infer metrics
+            if problem_property.fallback_validation_metric is not None:
```
[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:126]: Same as above. `SEMANTIC_SEGMENTATION`, `CLASSIFICATION`, matching, `FEATURE_EXTRACTION`, and `ZERO_SHOT_IMAGE_CLASSIFICATION` does not need fallback_validation_metric

File: multimodal/tests/unittests/others/test_metrics.py
Comment (FANGAreNotGnu) [reply multimodal/tests/unittests/others/test_metrics.py:1]: Sure I'll try to add one

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils/metric.py:119]: We can add IOU for semantic segmentation's evaluation and validation.

[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:119]: Sure, I'll add that

[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:119]: Do you mean binary_iou?

[the remark this answers is not in the document]
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:119]: Oh the name is changed.

File: multimodal/src/autogluon/multimodal/constants.py [hunk 7]
```diff
@@ -160,7 +169,12 @@
     BINARY_ACC: MAX,
     BER: MIN,
 }
-VALID_METRICS = METRIC_MODE_MAP.keys()
+
+MATCHING_METRICS = {
+    MULTICLASS: [SPEARMANR, SPEARMANR],
+    REGRESSION: [SPEARMANR, SPEARMANR],
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/constants.py:175]: `BINARY: [ROC_AUC, ROC_AUC],`  is missing.

File: multimodal/src/autogluon/multimodal/constants.py [hunk 8]
```diff
@@ -160,7 +169,12 @@
...
+}
+MATCHING_METRICS_WITHOUT_PROBLEM_TYPE = [RECALL, NDCG]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/constants.py:177]: recall should be the validation metric, but used as eval metric in `infer_metrics`.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 9]
```diff
@@ -74,62 +75,66 @@ def infer_metrics(
...
+                f"Metric {eval_metric_name} and metric {validation_metric_name} are used "
+                f"as the evaluation metric and the validation metric for matching tasks by default. "
+            )
             return validation_metric_name, eval_metric_name
-
-        # Currently only support recall as validation metric in retrieval tasks.
-        if is_matching and eval_metric_name in RETRIEVAL_METRICS:
-            validation_metric_name = RECALL
+        elif problem_type in MATCHING_METRICS:
+            eval_metric_name, validation_metric_name = MATCHING_METRICS[problem_type]
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:103]: Ditto. Switch the order of return.

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zhiqiangdon) [reply multimodal/src/autogluon/multimodal/utils/metric.py:128]: If invalid evaluation or validation metrics are provided and the default eval and validation metrics are None, then this function still return the invalid ones. Does it make sense?

File: multimodal/tests/unittests/others/test_metrics.py [hunk 10]
```diff
@@ -84,6 +85,78 @@ def test_bce_with_logits_loss(problem_type, loss_func_name):
...
+    val_score = predictor._learner._best_score
+    eval_score = predictor.evaluate(dataset.test_df)[eval_metric]
+    assert abs(val_score - eval_score) < 1e-4
+
+
+@pytest.mark.single_gpu
+@pytest.mark.parametrize(
+    "problem_type, eval_metric_name, validation_metric_name, is_matching, target_eval_metric_name, target_validation_metric_name",
+    [
+        ("binary", "acc", None, False, "acc", "acc"),
```
Comment (zhiqiangdon) [inline multimodal/tests/unittests/others/test_metrics.py:133]: Better to cover all the default metrics for all problem types, i.e., providing None and None and getting the expected returns.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 11]
```diff
@@ -76,70 +79,93 @@ def infer_metrics(
...
+        # Fallback based on problem type unless it's a customized metric
+        if problem_type is None:
+            validation_metric_name, fallback_eval_metric_name = MATCHING_METRICS_WITHOUT_PROBLEM_TYPE
+            logger.info(
+                f"Metric {eval_metric_name} and metric {validation_metric_name} are used "
+                f"as the evaluation metric and the validation metric for matching tasks by default. "
+            )
+        elif problem_type in MATCHING_METRICS:
+            validation_metric_name, fallback_eval_metric_name = MATCHING_METRICS[problem_type]
+            logger.info(
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:121]: Change all `logger.info` to `logger.debug` in this function.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 12]
```diff
@@ -76,70 +79,93 @@ def infer_metrics(
...
+                f"as the evaluation metric and the validation metric for {problem_type} in matching tasks by default. "
+            )
+        else:
+            raise NotImplementedError(f"Problem type: {problem_type} is not yet supported for matching!")
+        if not is_customized:
+            eval_metric_name = fallback_eval_metric_name
+        logger.info(
+            f"Metric {eval_metric_name} and metric {validation_metric_name} are used "
+            f"as the evaluation metric and the validation metric for {problem_type} in matching tasks by default. "
+        )
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:132]: This log info has duplicated three times.

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 13]
```diff
@@ -76,70 +79,93 @@ def infer_metrics(
...
+        elif problem_type in MATCHING_METRICS:
+            validation_metric_name, fallback_eval_metric_name = MATCHING_METRICS[problem_type]
+            logger.info(
+                f"Metric {eval_metric_name} and metric {validation_metric_name} are used "
+                f"as the evaluation metric and the validation metric for {problem_type} in matching tasks by default. "
+            )
+        else:
+            raise NotImplementedError(f"Problem type: {problem_type} is not yet supported for matching!")
+        if not is_customized:
+            eval_metric_name = fallback_eval_metric_name
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:128]: The logic looks weird. Since eval metric is not None when running there, no need to check is_customized. I think a better way is to just put `eval_metric_name` in the above returns instead of doing `eval_metric_name = fallback_eval_metric_name`

File: multimodal/src/autogluon/multimodal/utils/metric.py [hunk 14]
```diff
@@ -76,70 +79,93 @@ def infer_metrics(
...
+        # Infer evaluation metric
+        if eval_metric_name.lower() not in problem_property.supported_evaluation_metrics and not is_customized:
+            warnings.warn(
+                f"Metric {eval_metric_name} is not supported as the evaluation metric for {problem_type}. "
+                f"The evaluation metric is changed to {problem_property.fallback_evaluation_metric} by default."
+            )
+            if problem_property.fallback_evaluation_metric is not None:
+                eval_metric_name = problem_property.fallback_evaluation_metric
+            else:
+                return None, None
```
Comment (zhiqiangdon) [inline multimodal/src/autogluon/multimodal/utils/metric.py:145]: Instead of adding the `return None, None`, how just remove `if problem_property.fallback_evaluation_metric is not None:`?

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 11 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:121]: Shall we by default let users know the metric is changed?

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 14 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:145]: but then we still need to check if the eval_metric_name is None while inferring validation metric below

File: multimodal/src/autogluon/multimodal/utils/metric.py (hunk 13 above)
Comment (FANGAreNotGnu) [reply multimodal/src/autogluon/multimodal/utils/metric.py:128]: At this point, the eval_metric_name is either None or invalid, so we fallback the metric if it's not a customized metric. We do want to fallback a customized metric.

**Review by zhiqiangdon — APPROVED** [review]
LGTM!
